crd-generator/collector/src/main/java/io/fabric8/crd/generator/collector/JandexCustomResourceClassScanner.java

```diff
@@ -0,0 +1,156 @@
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Collects multiple Jandex index sources and allows to find Custom Resource class names in the
+ * resulting composite index by using {@link JandexCustomResourceClassScanner#findCustomResourceClasses()}.
+ * <p>
+ * If not overridden by {@link JandexCustomResourceClassScanner#forceIndex}, the implementation uses an
+ * existing Jandex index if found in the source. Otherwise, the index will be created on the fly.
+ * </p>
+ *
+ * @see Index
+ * @see JandexIndexer
+ * @see JandexUtils
+ */
+public class JandexCustomResourceClassScanner {
+
+  private final List<IndexView> indices = new LinkedList<>();
+  private final Set<File> filesToScan = new HashSet<>();
+
+  /**
+   * If <code>true</code>, indices will always be created even if an index exists at the source.
+   */
+  private boolean forceIndex = false;
+
+  public JandexCustomResourceClassScanner withForceIndex(boolean forceIndex) {
+    this.forceIndex = forceIndex;
+    return this;
+  }
+
+  public JandexCustomResourceClassScanner withIndex(IndexView... index) {
+    if (index != null) {
+      withIndices(Arrays.asList(index));
+    }
+    return this;
+  }
+
+  public JandexCustomResourceClassScanner withIndices(Collection<IndexView> indices) {
+    if (indices != null) {
+      indices.stream()
+          .filter(Objects::nonNull)
+          .forEach(this.indices::add);
+    }
+    return this;
+  }
+
+  public JandexCustomResourceClassScanner withFileToScan(File... files) {
+    if (files != null) {
+      withFilesToScan(Arrays.asList(files));
+    }
+    return this;
+  }
+
+  public JandexCustomResourceClassScanner withFilesToScan(Collection<File> files) {
+    if (files != null) {
+      files.stream()
+          .filter(Objects::nonNull)
+          .forEach(this.filesToScan::add);
+    }
+    return this;
+  }
+
+  /**
+   * Finds Custom Resource classes in the pre-configured scope.
+   *
+   * @return the Custom Resource classes
+   */
+  public Collection<String> findCustomResourceClasses() {
+    List<IndexView> actualIndices = new ArrayList<>(indices);
+    actualIndices.add(createBaseIndex());
+
+    IndexView index = JandexUtils.createIndex(actualIndices, filesToScan, forceIndex);
+
+    return findCustomResourceClasses(index).stream()
+        .map(ClassInfo::toString)
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Finds Custom Resource classes in an index.
+   *
+   * @param index the index
+   * @return the Custom Resource classes
+   *
+   * @see JandexCustomResourceClassScanner#createBaseIndex()
+   */
+  private List<ClassInfo> findCustomResourceClasses(IndexView index) {
+    // Only works if HasMetadata and all intermediate classes
+    // (like CustomResource) are included in given index.
+    return index.getAllKnownImplementors(HasMetadata.class)
```

Comment:
Why are we looking at all implementors of `HasMetadata` instead of `CustomResource`? Is it meant to also include native resources?

Reply:
I have noticed somewhere that you want to get rid of the CustomResource class. That's why I tried to implement it without referencing the `CustomResource` class. It also works if we use here the CustomResource class.

If in the future other interfaces / classes are introduced to indicate a CustomResource, this method must be ajusted.

Reply:
Makes sense. My concern here is that, depending on the index, this might return a lot of results.

Reply:
I have found the discussions:
https://github.com/fabric8io/kubernetes-client/issues/4164#issuecomment-1328718564

Reply:
> My concern here is that, depending on the index, this might return a lot of results.

If it's really a problem we can use CustomResource class for now until the replacement is in place. I have no problem to change it back within this PR.

Do you have some numbers about the biggest known projects?

Reply:
We can revisit if that becomes an issue, for sure. I don't really have a number but I'm thinking if people are importing a lot of Fabric8 libs and these get use in a Quarkus context, they might all end up in the application index, which should probably be used when generating CRDs from a Quarkus app, so there might be lots of candidates there.

Reply:
Now I understand your concerns better. Some clearifications:

**QOSDK Projects**

QOSDK projects won't use the collector module because QOSDK directly integrates with the CRD-Generator v1 and it seems like you want to integrate in the same way the CRD-Generator v2 in https://github.com/quarkiverse/quarkus-operator-sdk/pull/876, too. 
I think this is good and the way for the future, because Quarkus already provides features to search for classes  during build time.

In conclusion the recommendation for QOSDK projects to generate CRDs is at the moment and should be in the future to use the QOSDK native way.

**Quarkus Projects**

If a Quarkus project, which doesn't use QOSDK, wants to generate CRDs, the project might use either the new CRD-Generator Maven or Gradle plugin.
A Quarkus project already builds a Jandex index for the *project classes* and *submodules* automatically. This index contains external classes from libs only if *explicitly* configured e.g. by `@RegisterForReflection`.

The collector module will detect this index by default, skips building an own index and uses it to search for custom resources. Other indices from libraries are only considered *if explicitly configured*.  (e.g. in the maven plugin with `dependenciesToScan`).

> I'm thinking if people are importing a lot of Fabric8 libs and these get use in a Quarkus context, they might all end up in the application index, which should probably be used when generating CRDs from a Quarkus app, so there might be lots of candidates there.

In conclusion, the relevant index is in most projects smaller as you might have initially thought.
The overal recommendation would be to use features to extend the index (e.g. `dependenciesToScan`) wisely and only if necessary.


Coming back to the original topic, I think we should look at this question more on a strategic perspective.

**What interfaces / annotations / classes should be used within this PR to identify custom resource classes?**

Options:

1. A class is a custom resource class if it extends `CustomResource`
2. A class is a custom resource class if it implements `HasMetadata` and is annotated by `@Groupd` and `@Version` (current implementation)

@manusa @shawkins you might have also an opinion on that. What option is preferred?

Reply:
I'm not sure how we're excluding the core classes here (i.e. any of the model types that we generate as part of `kuberentes-model-generator` sub modules).
These classes implement HasMetadata and are annotated with Group and Version.

Reply:
> Is this even possible?

I'm not sure (because I don't know Quarkus in depth), but I don't think this is a problem.

I have started to write some docs and diagrams, so that we can better talk about this:
https://github.com/baloo42/crd-generator-v2-docs/blob/main/crd-generator-maven-plugin.md

The first example shows a default setup where custom resource classes are only located in the current project.
The second example shows a more advanced setup, where CRDs should be generated for custom resource classes in the project **and** for custom resource classes in a dependency.

In both examples, having an jandex index, is optional. But if it exist, it will be used and will therefore result in a small performance improvement.

If an index exists, it is assumed, that this index is not messy and contains only classes of the project (`target/**`) or of the dependency itself. (I'm a little bit unsure if we can assume this always).

> In any case, if that happens, the plugin should provide the option for inclusion/exclusion of classes/packages. This I understand it already does in the current implementation Inclusions, used to filter Custom Resource classes after scanning..
> However, if what metacosm describes is possible, the troublesome part would be the after scanning since the initial loading would require an increased overhead, so filtering should probably happen at this point.

At the moment there are two options to control inclusions/exclusions:

1. By defining the scope for scanning. (Which is something like "inclusions")
2. By filtering after the scan with package inclusions/exclusions.

To explain the first option more in detail, let's focus on the [second example in the draft docs](https://github.com/baloo42/crd-generator-v2-docs/blob/main/crd-generator-maven-plugin.md#scan-scope-project-and-dependency) and assume the "Jar Archive 1" is the kubernetes-client library (and all its dependencies / submodules). If I understand you both correct, your concerns for such a scenario are, that the list of found custom resource classes would be big.
I don't think this is the case, because as shown in the second example, only explicitly defined dependencies are used to build up the internal composite index. Therefore only custom resource classes of explicitly defined dependencies would show up in the list.

I also think, that even the second option is very efficient, because it is also processed before the class is loaded and handed over to the crd-generator.

In conclusion, I think a user won't get into troubles as long as the user creates conventional jandex indices.

The only thing I can imagine as a future improvement is that we allow a user to provide an index and allow to disable detecting existing indices as well as to disable creating indices automatically.
This would give the user full control over the internal index of the maven-plugin, so that he can optimize the scope.

---
Sidenotes:

Beside it makes sense or not, I have searched for a way to get classes from an index filtered by multiple criterias (like by package and by implementing interface). But I always ended up to stream the first results and filter further. I don't think there is a native way to do this with jandex.

I also explored if it's possible to create subindices with a subset of the classes of an index. But I haven't found an efficient way.

Reply:
>  only explicitly defined dependencies are used to build up the internal composite index

I think this resolves my concerns. Just to make sure, here I'm repeating what I get from your explanation (+ detailed diagrams and use/config cases) _Users will opt-in to scan their dependencies in their pom.xml plugin configuration, by adding these dependencies explicitly they should be aware of any potential performance issues created by huge model artifacts_

Reply:
> Users will opt-in to scan their dependencies in their pom.xml plugin configuration, by adding these dependencies explicitly they should be aware of any potential performance issues created by huge model artifacts

That's correct. In addition we should add the following warning:

> A serialized jandex index in the project (`target/classes/META-INF/jandex.idx` will be detected and used. Be aware of performance issues if this index contains (many) classes from dependencies. (This shouldn't be the case if you create a conventional jandex index by including the project files only).